File: app/src/main/java/com/OdiousPanda/rainbow/Utilities/ClothesIconUtil.java
package com.OdiousPanda.rainbow.Utilities;

import android.content.Context;

import com.OdiousPanda.rainbow.DataModel.Weather.Weather;
import com.OdiousPanda.rainbow.R;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class ClothesIconUtil {
    private Weather weather;
    private String criteria = "cool";
    private boolean rain = false;
    private List<Integer> head = new ArrayList<>();
    private List<Integer> upper = new ArrayList<>();
    private List<Integer> lower = new ArrayList<>();
    private List<Integer> foot = new ArrayList<>();
    private List<Integer> hand = new ArrayList<>();

    public ClothesIconUtil(Weather weather) {
        this.weather = weather;
        generateCriteria();
        getIcons();
    }

    private void getIcons() {
        Field[] drawablesFields = com.OdiousPanda.rainbow.R.drawable.class.getFields();
        for (Field field : drawablesFields) {
            try {
                if (field.getName().contains("head_")) {
                    if (field.getName().contains(criteria)) {
                        head.add(field.getInt(null));
                    }
                } else if (field.getName().contains("upper_")) {
                    if (field.getName().contains(criteria)) {
                        upper.add(field.getInt(null));
                    }
                } else if (field.getName().contains("lower_")) {
                    if (field.getName().contains(criteria)) {
                        lower.add(field.getInt(null));
                    }
                } else if (field.getName().contains("foot_")) {
                    if (field.getName().contains(criteria)) {
                        foot.add(field.getInt(null));
                    }
                } else if (field.getName().contains("hand_")) {
                    hand.add(field.getInt(null));
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private void generateCriteria() {
        float temp = UnitConverter.toCelsius(weather.getCurrently().getApparentTemperature());
        String summary = weather.getCurrently().getIcon();
        if (temp > 30) {
            criteria = "hot";
        } else if (temp < 15) {
            criteria = "cold";
        } else {
            criteria = "cool";
        }
        if (summary.contains("rain")) {
            rain = true;
        }
    }

    public String getCause(Context context) {
        String criteriaString = context.getString(R.string.cool);
        if (criteria.equals("hot")) {
            criteriaString = context.getString(R.string.hot);
        } else if (criteria.equals("cold")) {
            criteriaString = context.getString(R.string.cold);
        }
        return rain ? " " + criteriaString + " " + context.getString(R.string.and_raining) : " " + criteriaString + ".";
    }

    public int getHeadIcon() {
        return head.get(new Random().nextInt(head.size()));
    }

    public int getUpperIcon() {
        return upper.get(new Random().nextInt(upper.size()));
    }

    public int getLowerIcon() {
        return lower.get(new Random().nextInt(lower.size()));
    }

    public int getFootIcon() {
        return foot.get(new Random().nextInt(foot.size()));
    }

    public int getHandIcon() {
        if (rain) {
            return R.drawable.rain_ic_umbrella;
        }
        return hand.get(new Random().nextInt(hand.size()));
    }
}

File: app/src/main/java/com/odiousPanda/rainbowKt/utilities/ClothesIconUtil.kt
package com.odiousPanda.rainbowKt.utilities

import android.content.Context
import com.odiousPanda.rainbowKt.R
import com.odiousPanda.rainbowKt.model.dataSource.weather.Weather
import java.lang.reflect.Field
import java.util.*

class ClothesIconUtil {
    private var criteria = "cool"
    private var rain = false
    private val head = mutableListOf<Int>()
    private val upper = mutableListOf<Int>()
    private val lower = mutableListOf<Int>()
    private val foot = mutableListOf<Int>()
    private val hand = mutableListOf<Int>()

    init {
        getIcons()
    }

    private fun getIcons() {
        val drawablesFields: Array<Field> = R.drawable::class.java.fields

        for (field in drawablesFields) {
            try {
                if (field.name.contains("head_")) {
                    if (field.name.contains(criteria)) {
                        head.add(field.getInt(null))
                    }
                } else if (field.name.contains("upper_")) {
                    if (field.name.contains(criteria)) {
                        upper.add(field.getInt(null))
                    }
                } else if (field.name.contains("lower_")) {
                    if (field.name.contains(criteria)) {
                        lower.add(field.getInt(null))
                    }
                } else if (field.name.contains("foot_")) {
                    if (field.name.contains(criteria)) {
                        foot.add(field.getInt(null))
                    }
                } else if (field.name.contains("hand_")) {
                    hand.add(field.getInt(null))
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun generateCriteria(weather: Weather) {
        val temp =
            UnitConverter.toCelsius(weather.currently.apparentTemperature)
        rain = weather.currently.icon.contains("rain")
        criteria = when {
            temp > Constants.APT_TEMP_HOT -> {
                "hot"
            }
            temp < Constants.APT_TEMP_COLD -> {
                "cold"
            }
            else -> {
                "cool"
            }
        }
    }

    fun getCause(context: Context): String? {
        var criteriaString = context.getString(R.string.cool)
        if (criteria == "hot") {
            criteriaString = context.getString(R.string.hot)
        } else if (criteria == "cold") {
            criteriaString = context.getString(R.string.cold)
        }
        return context.resources.getString(R.string.cause_it_s) + if (rain) " $criteriaString ${context.getString(R.string.and_raining)}." else " $criteriaString."
    }

    fun getHeadIcon(): Int {
        return head[Random().nextInt(head.size)]
    }

    fun getUpperIcon(): Int {
        return upper[Random().nextInt(upper.size)]
    }

    fun getLowerIcon(): Int {
        return lower[Random().nextInt(lower.size)]
    }

    fun getFootIcon(): Int {
        return foot[Random().nextInt(foot.size)]
    }

    fun getHandIcon(): Int {
        return if (rain) {
            R.drawable.rain_ic_umbrella
        } else hand[Random().nextInt(hand.size)]
    }
}